bluesky_kafka/kafka.py

```diff
@@ -22,13 +36,20 @@ class Publisher:
 
     Reference: https://github.com/confluentinc/confluent-kafka-python/issues/137
 
+    The default configuration of the underlying Kafka Producer is an "idempotent"
+    producer. This means three things:
+        1) delivery acknowledgement is not sent until all replicate brokers have received a message
+        2) message delivery will be retried indefinitely (messages will not be dropped by the Producer)
+        3) message order will be maintained
+
     Parameters
     ----------
-    address : string
-        Address of a running Kafka server as a string like
-        ``'127.0.0.1:9092'``
+    bootstrap_servers : str
+        Comma-delimited list of Kafka server addresses as a string such as ``'127.0.0.1:9092'``
+    producer_config: dict, optional
```

Comment:
Since is Kafka-specific, can we make it the last argument? A small price to pay for API compatibility with the `bluesky.callbacks.zmq.Publisher`.

Reply:
I feel the better design to put Kafka-specific arguments _first_, and I continue my objection to making the 0MQ and Kafka Publishers API-compatible:
 - I don't think 0MQ and Kafka are similar enough that it is reasonable to write code that runs against both. For example, 0MQ is basically stateless while Kafka holds messages for a long time and consumers can request delivery of "old" messages or "new" messages (relative to when the consumer was created). On top of that the RemoteDispatchers can't be API compatible because Kafka Consumers require some configuration beyond what is necessary for 0MQ, in particular the "group_id" which the Kafka brokers use to distribute messages of the same topic among different consumers for higher throughput. In a nutshell if code works with 0MQ there is no reason to expect it will work the same way with Kafka.
 - The effort required to install and run Kafka is not a vaild argument for API compatibility either. Kafka is no more difficult to install and run on a development system than 0MQ. To make it easy one can use the same script as CI to build a docker image:`scripts/kafka-docker-compose.yml`. I think it is even simpler to just download and unpack Kafka although that requires a bit of editing config files. I recommend these instructions: https://www.digitalocean.com/community/tutorial_collections/79.

A motivated person could probably work out some way to make API-compatible Publishers, but I think that would be a flawed design and would ultimately cause problems. If there is _still_ a desire to make these classes API-compatible I feel it is a case of "we do what we must because we can" and I _still_ disagree and object.

Reply:
I'll concede that in early iterations of this discussion I overestimated the difficulty of getting Kafka up and running---the [quickstart](https://kafka.apache.org/quickstart) really is quite direct and minimal. Also, it not much more resource intensive. (And in any case, while the proverbial "laptop user on 4G on a train" is not a case I think we need to _optimize_ for. We just need to make sure the minimal use case for bluesky is lightweight, and it is.)

> I don't think 0MQ and Kafka are similar enough that it is reasonable to write code that runs against both. 

I agree with that only narrowly. Yes, it looks like they cannot be made drop-in replacements for one another, but I think there is good reason to make the API's "rhyme"---to make it as smooth as possible to move between them. 0MQ PUB-SUB and Kafka are both implementations of publish--subscribe. We, or our users, may add more support for other message buses in the future (RabbitMQ, ActiveMQ, etc.) It is a good API design and common practice to placed shared concepts first in a consistent order (address(es), topics, choice of serialization) and then implementation-specific details at the end of the signature.

So let's suppose we eventually decide to deprecate `bluesky-0MQ-proxy` in order to guide our users toward more robust, feature-complete message buses and shed the maintenance burden of maintaining our own 0MQ PUB-SUB wrapper CLI. I could imagine that, though I'm sure yet I'd advocate for it. I think in that scenario, this API design would be still be justified as a way of laying track for future message bus integrations with "rhyming" if not literally drop-in consistent APIs.

Reply:
I think track is laid for future message bus integration not in `Publisher.__init__()` signatures but rather with `Publisher.__call__()`. I think it is more valuable to design `__init__` to aid developers in setting up a *Kafka* Publisher correctly and not as an exercise, however intellectually satisfying, in abstract categorization of message bus parameters. I cite scikit-learn as an example of a well-regarded project with a uniform API (`fit()`, `predict()`, ...) and `__init__` signatures emphasizing the specifics of each operation (rather than prioritizing the generalities).

This is my proposal for a `Publisher.__init__` that tries to guide developers to a correct initialization. All but the last argument have the same names found in the Kafka documentation. The first three arguments are defining characteristics of a Kafka Producer followed by the critical (but not "characteristic") Kafka broker addresses and the optional dictionary allowing customization of the many, many details of a Producer[1]. The last is the only argument that is not specific to Kafka and additionally *should* have a default for bluesky.
```python
def __init__(
    topic,     # defining characteristic, no default 
    group_id,  # defining characteristic, no default
    key,       # defining characteristic, no default
    bootstrap_servers,  # locations of the Kafka brokers, no default
    producer_config=None, # optional configuration parameters
    serializer=msgpack.dumps  # necessary for a message bus but Kafka is unaware of it
)
``` 

[1] It has become clear to me that some additional configuration functions similar to `logging.basicSetup` would be valuable in keeping Kafka Publisher configurations for production and testing from getting mixed up.

Reply:
OK, I'm sold on your argument for parameter naming/ordering. I'm less sure about [1], wondering if that could be just as well served by documentation.